Framework/Algorithms/src/AverageSpectrumBackground.cpp

```diff
@@ -0,0 +1,151 @@
+  alg->setProperty("KeepUngroupedSpectra", false);
+  alg->execute();
+  return alg->getProperty("OutputWorkspace");
+}
+
+std::vector<size_t> AverageSpectrumBackground::getSpectraFromRange(
+    const std::vector<size_t> range) {
+  const auto start = range[0];
+  const auto end = range[1];
+  std::vector<size_t> spectra;
+  for (auto itr = start; itr < end; ++itr) {
+    spectra.push_back(itr);
+  }
+  return spectra;
+}
+
+//----------------------------------------------------------------------------------------------
+/** Initialize the algorithm's properties.
+ */
+void AverageSpectrumBackground::init() {
+
+  // Input workspace
+  declareProperty(
+      Kernel::make_unique<API::WorkspaceProperty<API::MatrixWorkspace>>(
+          "InputWorkspace", "", Kernel::Direction::Input,
+          boost::make_shared<API::CommonBinsValidator>()),
+      "An input workspace.");
+
+  // bottom background range
+  declareProperty(Kernel::make_unique<Kernel::ArrayProperty<size_t>>(
+                      "BottomBackgroundRange", std::vector<size_t>()),
+                  "A list of the bottom background ranges.");
+  // top background range
+  declareProperty(Kernel::make_unique<Kernel::ArrayProperty<size_t>>(
+                      "TopBackgroundRange", std::vector<size_t>()),
+                  "A list of the top background ranges.");
+
+  // Output workspace
+  declareProperty(Kernel::make_unique<API::WorkspaceProperty<>>(
+                      "OutputWorkspace", "", Kernel::Direction::Output),
+                  "A Workspace with the background removed.");
+}
+
+//----------------------------------------------------------------------------------------------
+/** Execute the algorithm.
+ */
+void AverageSpectrumBackground::exec() {
+  API::MatrixWorkspace_sptr inputWS = getProperty("InputWorkspace");
+  std::vector<size_t> bottomBgdRange = getProperty("BottomBackgroundRange");
+  std::vector<size_t> topBgdRange = getProperty("TopBackgroundRange");
+
+  if (bottomBgdRange.empty() && topBgdRange.empty())
+    throw std::invalid_argument("At least one background range is required");
+  if (bottomBgdRange.size() != 2 && !bottomBgdRange.empty())
+    throw std::invalid_argument("BottomBackgroundRange must have length 2");
+  if (topBgdRange.size() != 2 && !topBgdRange.empty())
+    throw std::invalid_argument("TopBackgroundRange must have length 2");
+  if ((!bottomBgdRange.empty() && bottomBgdRange[1] - bottomBgdRange[0] == 0) ||
+      (!topBgdRange.empty() && topBgdRange[1] - topBgdRange[0] == 0)) {
+    throw std::invalid_argument("Cannot have a range of length 0");
+  }
+
+  API::MatrixWorkspace_sptr bottomBgd, topBgd;
+  // groups the bottom background into a single spectra workspace
+  if (!bottomBgdRange.empty()) {
+    bottomBgd =
+        groupBackgroundDetectors(inputWS, getSpectraFromRange(bottomBgdRange));
+  }
+
+  // groups the top background into a single spectra workspace
+  if (!topBgdRange.empty()) {
+    topBgd =
```

Comment:
Would it be possible to call GroupDetectors once with the combined list of input spectra, since we're summing the results anyway?